lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
+ */
+DataAccessObject.patchOrCreateWithWhere =
+DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) { return connectionPromise; }
+  assert(arguments.length >= 1, 'At least one argument is required');
+  if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+    data = where;
+    where = {};
+  } else if (options === undefined && cb === undefined) {
+        // One of:
+        // updateAll(data, cb)
+        // updateAll(where, data) -> Promise
+    if (typeof data === 'function') {
+      cb = data;
+      data = where;
+      where = {};
+    }
+  } else if (cb === undefined) {
+        // One of:
+        // updateAll(where, data, options) -> Promise
+        // updateAll(where, data, cb)
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+  data = data || {};
+  options = options || {};
+  cb = cb || utils.createPromiseCallback();
+  assert(typeof where === 'object', 'The where argument must be an object');
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+  var hookState = {};
+  var filter = { 'where': where };
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+  var modelObject = connector._models;
+  var modelName = Object.keys(modelObject)[0];
+  var id = getIdValue(this, data);
+  var context = {
+    Model: Model,
+    query: { 'where': where },
+    hookState: hookState,
+    options: options,
+  };
```

Comment:
Please preserve consistency with other existing operation hooks. There is no `filter` context property. The `access` hook provides `ctx.query`, `before save` provides `ctx.where`, `ctx.data`, `ctx.currentInstance` in case of a partial update, and `before save` provides `ctx.instance`.

Ideally, the operation hooks invoked for `upsertWithWhere` should be the same regardless of whether an atomic connector-provided function is called, or whether a non-atomic `find` + `updateAttributes` is used.

Take a look at [support/describe-operation-hooks.js](https://github.com/strongloop/loopback-datasource-juggler/blob/cca14a0230e51cb47795742f903e0eec7f35d7dd/support/describe-operation-hooks.js), it creates a comprehensive table showing what context is provided for different DAO operations.